sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -971,22 +977,27 @@ public void teardown() throws Exception {
 
     @ProcessElement
     public void processElement(ProcessContext c) throws Exception {
-      Sleeper sleeper = Sleeper.DEFAULT;
-      BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
-
-      Iterable<Mutation> mutations = c.element();
-
-      while (true) {
-        // Batch upsert rows.
-        try {
-          spannerAccessor.getDatabaseClient().writeAtLeastOnce(mutations);
-          // Break if the commit threw no exception.
-          break;
-        } catch (AbortedException exception) {
-          // Only log the code and message for potentially-transient errors. The entire exception
-          // will be propagated upon the last retry.
-          if (!BackOffUtils.next(sleeper, backoff)) {
-            throw exception;
+      Iterable<MutationGroup> mutations = c.element();
+      boolean tryIndividual = false;
+      // Batch upsert rows.
+      try {
+        Iterable<Mutation> batch = Iterables.concat(mutations);
+        spannerAccessor.getDatabaseClient().writeAtLeastOnce(batch);
+      } catch (SpannerException e) {
+        if (failureMode == FailureMode.REPORT_FAILURES) {
+          tryIndividual = true;
+        } else if (failureMode == FailureMode.FAIL_FAST) {
+          throw e;
+        } else {
+          throw new IllegalArgumentException("Unknown failure mode " + failureMode);
+        }
+      }
+      if (tryIndividual) {
+        for (MutationGroup mg : mutations) {
+          try {
+            spannerAccessor.getDatabaseClient().writeAtLeastOnce(mg);
+          } catch (SpannerException e) {
+            c.output(mg);
```

Comment:
Would be good to add some logging to this function. If I got a bunch of failed mutations, I'd appreciate knowing why they failed. I don't know if it makes sense to embed the failure information into each failed mutation, it might.

Reply:
Done